### PR TITLE
feat: [TW-7] shared table component and hooks

### PR DESCRIPTION
**Pull request by spartan-vietphan**
## Summary

- implement shared table component and hooks
- basic usage at user view component
- define type for user list example

## Types of Changes

- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which does not change existing behavior or add new functionality)
- [ ] Library update (non-breaking change that will update one or more libraries to newer versions)
- [ ] Domestic (documentation, non-breaking change that doesn't change code behavior, can skip testing)

## Test Plan:

Dev test:

https://github.com/c0x12c/spartan-template-react/assets/122691615/a68192ce-b64a-42be-b5f2-af48d9de74c1

## Jira Issues:

https://c0x12c.atlassian.net/browse/TW-7

### CODE REVIEW

File: src/utils/table.ts [hunk 1]
```diff
@@ -65,3 +71,8 @@ export function applyFilter<T extends { name: string }>({
 
   return inputData;
 }
+
+const SORT_ORDER_ORDER: SortOrder[] = ['desc', 'asc', undefined, 'desc'];
```
Comment (spartan-bachdkle) [inline src/utils/table.ts:75]: @spartan-vietphan please remove redundant `desc` item

File: src/shared/hooks/useTableSelection.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import { useCallback, useEffect, useMemo, useState } from 'react';
+
+import { TableItem } from '~shared/types/table.ts';
```
Comment (spartan-bachdkle) [inline src/shared/hooks/useTableSelection.ts:3]: Update `index.ts` for `shared/types` folder. We should import `import { TableItem } from '~shared' folder instead

File: src/configs/user.ts [hunk 3]
```diff
@@ -3,23 +3,38 @@ import { sample } from 'lodash';
...
+  UxDesigner = 'UX Designer',
+  UiUxDesigner = 'UI/UX Designer',
+  ProjectManager = 'Project Manager',
+  BackendDeveloper = 'Backend Developer',
+  FullStackDesigner = 'Full Stack Designer',
+  FrontEndDeveloper = 'Front End Developer',
+  FullStackDeveloper = 'Full Stack Developer',
+}
+
+export interface User {
```
Comment (spartan-bachdkle) [inline src/configs/user.ts:23]: We should follow convention `IUser` for interface, prefix with `I`.

File: src/modules/user/components/UserView.tsx [hunk 4]
```diff
@@ -1,81 +1,31 @@
...
-  UserTableRow,
-  UserTableToolbar,
-} from './molecules';
+import { UserTableToolbar } from './molecules';
 
-import { users } from '~configs';
+import { User, users } from '~configs';
+import { USER_VIEW_TABLE_COLUMNS } from '~modules/user/constants/user-view.tsx';
 import { Iconify, Scrollbar } from '~shared';
+import { Table } from '~shared/components/molecules/table';
```
Comment (spartan-bachdkle) [inline src/modules/user/components/UserView.tsx:15]: We should change the import to `import { Table } from '~shared';`

File: src/modules/user/components/UserView.tsx [hunk 5]
```diff
@@ -1,81 +1,31 @@
...
+import useTableSelection from '~shared/hooks/useTableSelection.ts';
```
Comment (spartan-bachdkle) [inline src/modules/user/components/UserView.tsx:16]: We should change the import to `import { useTableSelection } from '~shared'`;

File: src/modules/user/components/UserView.tsx [hunk 6]
```diff
@@ -1,81 +1,31 @@
...
-  TableNoData,
-  UserTableHead,
-  UserTableRow,
-  UserTableToolbar,
-} from './molecules';
+import { UserTableToolbar } from './molecules';
 
-import { users } from '~configs';
+import { User, users } from '~configs';
+import { USER_VIEW_TABLE_COLUMNS } from '~modules/user/constants/user-view.tsx';
```
Comment (spartan-bachdkle) [inline src/modules/user/components/UserView.tsx:13]: We should change the `import { USER_VIEW_TABLE_COLUMNS } from`'~modules/user'` or `import { USER_VIEW_TABLE_COLUMNS } from`'~modules'`

Comment (spartan-vietphan) [reply src/modules/user/components/UserView.tsx:13]: I think I should be
`import { USER_VIEW_TABLE_COLUMNS } from '../constants';`

File: src/shared/components/molecules/table/CustomTableBody.tsx [hunk 7]
```diff
@@ -0,0 +1,56 @@
+import TableBody from '@mui/material/TableBody';
+
+import CustomTableRow from '~shared/components/molecules/table/CustomTableRow.tsx';
```
Comment (spartan-bachdkle) [inline src/shared/components/molecules/table/CustomTableBody.tsx:3]: @spartan-vietphan please fix here

File: src/shared/components/molecules/table/CustomTable.tsx [hunk 8]
```diff
@@ -0,0 +1,71 @@
+import { CheckboxProps } from '@mui/material/Checkbox';
+import Table from '@mui/material/Table';
+import TableContainer from '@mui/material/TableContainer';
+
+import CustomTableBody from '~shared/components/molecules/table/CustomTableBody.tsx';
```
Comment (spartan-bachdkle) [inline src/shared/components/molecules/table/CustomTable.tsx:5]: @spartan-vietphan please fix here

File: src/shared/components/molecules/table/CustomTableHead.tsx [hunk 9]
```diff
@@ -0,0 +1,79 @@
+import Box from '@mui/material/Box';
+import Checkbox, { CheckboxProps } from '@mui/material/Checkbox';
+import TableCell from '@mui/material/TableCell';
+import TableHead from '@mui/material/TableHead';
+import TableRow from '@mui/material/TableRow';
+import TableSortLabel from '@mui/material/TableSortLabel';
+
+import { SortOrder, TableColumn, TableItem } from '~shared/types/table.ts';
```
Comment (spartan-bachdkle) [inline src/shared/components/molecules/table/CustomTableHead.tsx:8]: @spartan-vietphan fix here too

File: src/shared/types/table.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
+import { ReactNode } from 'react';
+import { TableCellProps } from '@mui/material/TableCell';
+
+export type SortOrder = 'desc' | 'asc' | undefined;
+
+export interface SortInfo {
```
Comment (spartan-bachdkle) [inline src/shared/types/table.ts:6]: @spartan-vietphan please change the interface and type follow naming variable conventions

**Review by spartan-bachdkle — COMMENTED** [review]
I left some comments.

**Pull request merged**